src/Http/Http.Abstractions/src/Internal/PathStringHelper.cs

```diff
@@ -24,24 +27,57 @@ internal class PathStringHelper
             true,  true,  true,  false, false, false, true,  false,     // 0x78 - 0x7F
         };
 
+        // Map from an ASCII char to its hex value, e.g. arr['b'] == 11. 0xFF means it's not a hex digit.
+        // From coreclr/src/System.Private.CoreLib/shared/System/Number.Parsing.cs
+        // byte[] rather than ReadOnlySpan<byte> to avoid double .Length check https://github.com/dotnet/coreclr/issues/24209
+        private static readonly byte[] CharToHexLookup = new byte[]
```

Comment:
You can reduce the size of the lut if only lowercase is handled.
And for making the char lowercase just set the 5th bit `c |= (char)(1 << 5)` (or use 0x20 directly).
No check for `A-Z` should be needed, as all other value will result in `0xFF`.

So the lut may fit into a cache line.

Same can be applied for `ValidPathChars`.

Reply:
An idea...

The lut could be ommited, and the following code used
```c#
[MethodImpl(MethodImplOptions.NoInlining)]
private static bool AreFollowingTwoCharsHex(string str, int index)
{
    Debug.Assert(index < str.Length - 2);

    int i1 = str[index + 1];
    int i2 = str[index + 2];

    return IsHexadecimalChar(i1) && IsHexadecimalChar(i2);
}

[MethodImpl(MethodImplOptions.AggressiveInlining)]
private static bool IsHexadecimalChar(int c)
{
    return (byte)(c - 48) <= 9 || (byte)((c & ~0x20) - 65) <= 5;
}
```
`IsHexadecimalChar` was there before, but this version is better optimized.

<details>
   <summary>codegen</summary>

```asm
; Assembly listing for method PathStringHelper:AreFollowingTwoCharsHex(ref,int):bool
; Emitting BLENDED_CODE for X64 CPU with AVX - Unix
; optimized code
; rbp based frame
; partially interruptible
; Final local variable assignments
;
;  V00 arg0         [V00,T00] (  5,  5   )     ref  ->  rdi         class-hnd
;  V01 arg1         [V01,T01] (  4,  4   )     int  ->  rsi
;  V02 loc0         [V02,T06] (  3,  2   )     int  ->  rdi
;# V03 OutArgs      [V03    ] (  1,  1   )  lclBlk ( 0) [rsp+0x00]   "OutgoingArgSpace"
;  V04 tmp1         [V04,T02] (  3,  5   )     int  ->  rax         "impAppendStmt"
;  V05 tmp2         [V05,T07] (  2,  1   )    bool  ->  rax         "Inline return value spill temp"
;  V06 tmp3         [V06,T08] (  3,  1.50)    bool  ->  rax         "Inline return value spill temp"
;  V07 cse0         [V07,T03] (  3,  3   )     int  ->  rax         "ValNumCSE"
;  V08 cse1         [V08,T04] (  3,  3   )     int  ->  rdx         "ValNumCSE"
;  V09 cse2         [V09,T05] (  3,  3   )     int  ->  rsi         "ValNumCSE"
;
; Lcl frame size = 0

G_M2516_IG01:
       55                   push     rbp
       488BEC               mov      rbp, rsp
       90                   nop

G_M2516_IG02:
       8D4601               lea      eax, [rsi+1]
       8B5708               mov      edx, dword ptr [rdi+8]
       3BC2                 cmp      eax, edx
       7364                 jae      SHORT G_M2516_IG08
       4863C0               movsxd   rax, eax
       0FB744470C           movzx    rax, word  ptr [rdi+2*rax+12]
       83C602               add      esi, 2
       3BF2                 cmp      esi, edx
       7355                 jae      SHORT G_M2516_IG08
       4863F6               movsxd   rsi, esi
       0FB77C770C           movzx    rdi, word  ptr [rdi+2*rsi+12]
       8D70D0               lea      esi, [rax-48]
       400FB6F6             movzx    rsi, sil
       83FE09               cmp      esi, 9
       7E16                 jle      SHORT G_M2516_IG03
       83E0DF               and      eax, -33
       83C0BF               add      eax, -65
       0FB6C0               movzx    rax, al
       83F805               cmp      eax, 5
       0F9EC0               setle    al
       0FB6C0               movzx    rax, al
       85C0                 test     eax, eax
       7427                 je       SHORT G_M2516_IG06

G_M2516_IG03:
       8D47D0               lea      eax, [rdi-48]
       0FB6C0               movzx    rax, al
       83F809               cmp      eax, 9
       7E15                 jle      SHORT G_M2516_IG04
       83E7DF               and      edi, -33
       83C7BF               add      edi, -65
       400FB6C7             movzx    rax, dil
       83F805               cmp      eax, 5
       0F9EC0               setle    al
       0FB6C0               movzx    rax, al
       EB05                 jmp      SHORT G_M2516_IG05

G_M2516_IG04:
       B801000000           mov      eax, 1

G_M2516_IG05:
       5D                   pop      rbp
       C3                   ret

G_M2516_IG06:
       33C0                 xor      eax, eax

G_M2516_IG07:
       5D                   pop      rbp
       C3                   ret

G_M2516_IG08:
       E81899B878           call     CORINFO_HELP_RNGCHKFAIL
       CC                   int3

; Total bytes of code 121, prolog size 5 for method PathStringHelper:AreFollowingTwoCharsHex(ref,int):bool
; ============================================================
```
</details>

But I don't know if this is better than the lut-variant. It depends on the input-string and if the lut is already in the cache, etc. Benchmark-results may vary.

So far I couldn't come up with a way where `i1` and `i2` are checked as "packed chars" (i.e. `(i1 << 16) | i2`) at once.

Reply:
Thank you, resolved.